### ISSUE TITLE
Is the Haswell/Ivybridge fp64 support any closer to being merged?

### ISSUE THREAD

**Issue by axfelix** [Igalia/mesa#3]
Just curious -- feel like I heard about this months ago but there hasn't been any activity upstream. Now that stencil texturing on Haswell has been merged I think this is the only blocker for going from 3.3 -> 4.5...

**Comment by seanlaguna** [comment]
Also curious about IVB fp64. It looks like there's been a commit in some WIP branches to expose 4.0 (https://github.com/Igalia/mesa/commit/7b21c8b4a56f088eca09a18c0188380bd6f59bc5). Is this ready to test? Can I pull this commit and try to build Mesa with any possibility of success or stability?

**Comment by itoral** [comment]
We sent the first version of the Haswell patches about 2 months ago, we got review feedback that we have to address and we are working on it. Also, there have been some other changes going into the driver that we have to adapt to and this being a 100+ patch series it takes a while to do all this. I think we should be able to send another version of the series soon.

As for IvyBridge, yes, we are working on it and we have a branch already available that is mostly functional. It requires some polishing and cleaning-up and there are some limitations related to register spilling that might make some shader programs fail to compile, but otherwise it is acceptable. I have not been directly involved with the IvyBridge work though so I am not sure if the current status might be stable enough, I'll let others comment on that.

**Comment by samuelig** [comment]
@seanlaguna, I'm working on the IVB support together with @jasuarez. As @itoral said, this works is mostly done but we still need to work on cleaning up the branch and polishing it for upstream submission. There are limitations with register spilling that makes some shader programs fail to compile, so this might happen to you.

If you want to test it, I recommend you to use the branch called i965-fp64-gen7-ivb-scalar-vec4 but knowing that it is in beta-ish state. However, I expect it to be more or less stable (i.e. no GPU hangs and most things working fine modulo bugs).

**Comment by seanlaguna** [comment]
I should use that precise branch and not the rebase/backup branch?

**Comment by axfelix** [comment]
Thanks for the update! Hope my comment didn't seem too demanding, obviously there's some interest here :)

**Comment by itoral** [comment]
@seanlaguna: yes, that branch (for ivb), the others are temporary branches that we produce in the process of updating this branch.

@axfelix: not at all, it is great to see interest in this :)

**Comment by seanlaguna** [comment]
@samuelig @itoral Wow, just compiled and installed that branch, and it seems to be working fine! Thanks a bunch for your work on this! Any idea when this will be merged into Mesa officially?

**Comment by itoral** [comment]
@seanlaguna: I suppose it could take at least 1-2 months if we don't identify any important issues during review. The Haswell and IvyBridge implementations are kinda tricky, and there is some level of infrastructure missing in the vec4 backend that needs to be added for this to work, so it would not be surprising if we need to go through a few review iterations before we merge this.

**Comment by seanlaguna** [comment]
Am I correct that the ivb branch previously referenced just got rebased? Sorry, it's a bit hard to follow exactly how the Mesa branches interoperate!

Also, I'm quite interested in this work, please let me know if there's any help I can provide (perhaps there are some low-hanging fruits to be plucked).

**Comment by seanlaguna** [comment]
For what it's worth, right now the `i965-fp64-gen7-ivb-scalar-vec4` (I know, I know, I'm getting antsy...) fails on my machine.

**Comment by samuelig** [comment]
@seanlaguna yes, we have just rebased it. Any time we update the branch is because it got rebased and/or we added more patches to it :-)

We are very close to send it for review (we need to wait until HSW patches are pushed to upstream's master branch because we depend on them) and there isn't work. However, if we find a development task that you could do, we will notify you. Thanks for volunteering!

Regarding the failure, can you explain in which case it fails? Is it reproducible with piglit (https://piglit.freedesktop.org/) or is it running a game/application? In case of the latter, Can you give details about it (ideally giving us the source code of it) or give us a trace with apitrace (https://apitrace.github.io/), so we can reproduce the failure locally?

Thanks!

**Comment by seanlaguna** [comment]
Sorry, I edited my comment to be a bit more clear, but it seems to be too late. Tha failure happened as part of the `make check` tests. This actually fails on both branches now (`fp64` and `va64`), so it may not have to do with your work, but the rebase itself. It's the test written above, `test_vec4_register_coalesce`, here's my test log:

https://gist.github.com/seanlaguna/66ed242a211dc0f1f8e42edd71f47344

Not sure if this helps. I will install my build anyway and run piglit over it too if that helps.

**Comment by samuelig** [comment]
OK, thanks for the info! It should be enough to debug it! :-) Tomorrow is bank day here in Spain, so I will debug it on Thursday. Thanks!

**Comment by samuelig** [comment]
@seanlaguna We have found the source of the problem, thanks for the test log! We will push a patch fixing it to the branch once it is done. Thanks!

**Comment by seanlaguna** [comment]
Hey guys, I've been using the `i965-va64-gen7-ivb-scalar-vec4` branch for a while now. I know may not be terribly informative, but I've seen some log messages while running Chrome that indicate some possible bugs. Chrome on the whole has been working fine, so these aren't certain bugs, but might be interesting to look at:

https://gist.github.com/seanlaguna/6497abb3d3649f3e4094625b97744c35

**Comment by seanlaguna** [comment]
Any updates on this getting merged? Also, any reason why the 4.2 patches were removed from the va64 branch?

**Comment by samuelig** [comment]
Haswell patches are still on review and we need them in master before sending Ivybridge patches. Regarding the 4.2 patches, we sent a new patch for Haswell for review that fixes errors in some combination of kernel and HW. If it is accepted we will port it to the IVB branch. We have remove it meanwhile in the IVB VA64 branch, so we don't forget about it as the review is taking some time. If you need it, we can push the old patch to the branch.

**Comment by samuelig** [comment]
Edited previous comment to say that there was an error in our branch, it should be fine now.

**Comment by suberimakuri** [comment]
@seanlaguna 
In case you haven't seen, looks like the series is seeing reviews on the mesa-dev list. Mostly done, few minor changes/suggestions and one outstanding question with a previous author (Curro?). 
Getting closer!
https://lists.freedesktop.org/archives/mesa-dev/2016-December/137398.html

(I have HSW laptop and desktop so also looking forward to it being completed)

**Comment by seanlaguna** [comment]
Awesome. I am still compiling manually and everything has been working great. Definitely still following this.

**Comment by axfelix** [comment]
Congrats on getting this merged! I assume the 4.1 vertex attrib extension that's blocking Haswell from 4.5 isn't far behind? :)

**Comment by jasuarez** [comment]
On Thu, 2017-01-05 at 08:41 -0800, axfelix wrote:
> Congrats on getting this merged! I assume the 4.1 vertex attrib
> extension that's blocking Haswell from 4.5 isn't far behind? :)
> 
> 
> —

Right. We hope to send that extension for review in the following days.

**Comment by axfelix** [comment]
Awesome -- thanks for the great work, guys! By the next mesa release haswell will officially be better-supported on Linux than Windows or OSX :)

**Comment by seanlaguna** [comment]
Any word on Ivy Bridge? Is there an up-to-date branch with the vertex attrib patches for IVB I can pull? Hope this isn't too pushy...

**Comment by itoral** [comment]
@seanlaguna: We have just sent IVB/fp64 patches for review. As for vertex attrib, Juan has had patches for IVB vertex attrib ready for quite some time, but we need to wait for IVB/fp64 to land first. It should not take too long to merge vertex attrib after we land fp64.

I figure these are the branches with the latest code, but Juan should know better:
ivb/fp64:  i965-fp64-gen7-ivb-scalar-vec4-rc1
ivb/va64: jasuarez/i965-va64-gen7-ivb-scalar-vec4-20170104

**Comment by jasuarez** [comment]
> I figure these are the branches with the latest code, but Juan should
> know better:
> ivb/fp64: i965-fp64-gen7-ivb-scalar-vec4-rc1
> ivb/va64: jasuarez/i965-va64-gen7-ivb-scalar-vec4-20170104 


Correct. The first one adds support for fp64 on Ivy, and it is under
review.

The other one adds also support to va64 on Ivy. While it is fully
functional it will require some changes before sending to review.

**Comment by seanlaguna** [comment]
Sweet, thank you.

**Comment by samuelig** [comment]
Patches were pushed upstream. Close this issue.
